packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,140 @@
+import type { AddApprovalRequest } from '@metamask/approval-controller';
+import type { ControllerMessenger } from '@metamask/base-controller';
+import type { InfuraNetworkType } from '@metamask/controller-utils';
+import {
+  ApprovalType,
+  BUILT_IN_NETWORKS,
+  isNetworkType,
+} from '@metamask/controller-utils';
+import type { JsonRpcMiddleware } from '@metamask/json-rpc-engine';
+import { createAsyncMiddleware } from '@metamask/json-rpc-engine';
+import type {
+  NetworkClientId,
+  NetworkControllerFindNetworkClientIdByChainIdAction,
+  NetworkControllerGetNetworkClientByIdAction,
+  NetworkControllerGetStateAction,
+  NetworkControllerSetActiveNetworkAction,
+  NetworkControllerSetProviderTypeAction,
+} from '@metamask/network-controller';
+import type { SelectedNetworkControllerSetNetworkClientIdForDomainAction } from '@metamask/selected-network-controller';
+import { SelectedNetworkControllerActionTypes } from '@metamask/selected-network-controller';
+
+import type { QueuedRequestControllerEnqueueRequestAction } from './QueuedRequestController';
+import { QueuedRequestControllerActionTypes } from './QueuedRequestController';
+
+const isConfirmationMethod = (method: string) => {
+  const confirmationMethods = [
+    'eth_sendTransaction',
+    'wallet_watchAsset',
+    'wallet_switchEthereumChain',
+    'eth_signTypedData_v4',
+  ];
+
+  return !confirmationMethods.includes(method);
+};
+
+export const createQueuedRequestMiddleware = (
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >,
+  useRequestQueue: () => boolean,
+): JsonRpcMiddleware<any, any> => {
+  return createAsyncMiddleware(async (req: any, res: any, next) => {
+    if (!useRequestQueue() || isConfirmationMethod(req.method)) {
+      next();
+      return;
+    }
+
+    const networkClientIdForRequest = req.networkClientId as NetworkClientId;
```

Comment:
Would it be feasible to enforce that `networkClientId` come in under `req.params.networkClientId`? Not conforming to the middleware typing seems to be a big reason why `any` is being used so much here.

Reply:
Here are the proposed changes in a single commit for reference and testing: https://github.com/MetaMask/core/commit/6f9cfbf17fb80c3af8250915857f0057c3108b36

Reply:
`origin` is being added in the clients. In the extension, it's added [here](https://github.com/MetaMask/metamask-extension/blob/f0d81dcc514c92be1f5192b8af701a9ade35fcdf/app/scripts/lib/createOriginMiddleware.js), and in the mobile app, it's added [here](https://github.com/MetaMask/metamask-mobile/blob/c9a1b3e0d87e8344d6c17cd7a28cf8945d8b0c8b/app/util/middlewares.js#L29).

`networkClientId` is added by SelectedNetworkController [here](https://github.com/MetaMask/core/blob/a2d19d0713aac58b2e76b6dbdd31fcab92743d6a/packages/selected-network-controller/src/SelectedNetworkMiddleware.ts#L46).

In both cases, we control when this property is added. However, we're also assuming that those middleware are somewhere in the middleware stack and were executed before this one was. While this makes sense, I think we ought to explicitly look for these properties before attempting to use them. I will make some suggestions on how we can improve this.